scylla/src/transport/topology.rs

```diff
@@ -293,20 +385,297 @@ async fn query_keyspaces(conn: &Connection) -> Result<HashMap<String, Keyspace>,
+            clustering_key[position as usize] = column_name;
+        }
+
+        result.insert(
+            (keyspace_name.clone(), table_name.clone()),
+            Table {
+                columns,
+                partition_key,
+                clustering_key,
+                partitioner: partitioners
+                    .remove(&(keyspace_name, table_name))
+                    .unwrap_or_default(),
+            },
+        );
+    }
+
+    Ok(result)
+}
+
+fn map_string_to_cql_type(type_: &str) -> Result<CqlType, InvalidCqlType> {
+    use CollectionType::*;
+
+    let frozen = type_.starts_with("frozen<");
+
+    let type_ = if frozen {
+        let langle_position = type_.find('<').unwrap();
+        &type_[langle_position + 1..type_.len() - 1]
+    } else {
+        type_
+    };
+
+    if type_.ends_with('>') && type_.contains('<') {
+        let langle_position = type_.find('<').unwrap();
+        let inner_type = &type_[langle_position + 1..type_.len() - 1];
+        let outer_type = &type_[..langle_position];
+        let type_ = match outer_type {
+            "map" | "tuple" => {
+                let types = inner_type.split(", ").collect::<Vec<_>>();
+                if outer_type == "map" {
+                    if types.len() != 2 {
+                        return Err(InvalidCqlType(type_.to_string()));
+                    }
+                    Map(
+                        Box::new(map_string_to_cql_type(types[0])?),
+                        Box::new(map_string_to_cql_type(types[1])?),
+                    )
+                } else {
+                    return Ok(CqlType::Tuple(
+                        types
+                            .iter()
+                            .map(|type_| map_string_to_cql_type(type_))
+                            .collect::<Result<Vec<_>, _>>()?,
+                    ));
+                }
+            }
+            "set" => Set(Box::new(map_string_to_cql_type(inner_type)?)),
+            "list" => List(Box::new(map_string_to_cql_type(inner_type)?)),
+            _ => {
+                return Err(InvalidCqlType(type_.to_string()));
+            }
+        };
+        return Ok(CqlType::Collection { frozen, type_ });
+    }
+
+    Ok(match NativeType::from_str(type_) {
+        Ok(type_) => CqlType::Native(type_),
+        _ => CqlType::UserDefinedType {
+            frozen,
+            name: type_.to_string(),
+        },
+    })
+}
+
+async fn query_table_partitioners(
+    conn: &Connection,
+) -> Result<HashMap<(String, String), Option<String>>, QueryError> {
+    let mut partitioner_query = Query::new(
+        "select keyspace_name, table_name, partitioner from system_schema.scylla_tables",
+    );
+    partitioner_query.set_page_size(1024);
+
+    let rows = match conn.query_all(&partitioner_query, &[]).await {
+        Err(QueryError::DbError(DbError::Invalid, message))
+            if message == "table scylla_tables does not exist" =>
```

Comment:
I'd rather not rely on the error string. Different Cassandra versions might have some differences in the error message (e.g. spelling, capitalization, quotes, etc.). I understand that you would like this to match on the "table scylla_tables does not exist" error only, but currently I think checking the error code is the best thing we can do without making this check fragile. Please remove the string comparison.

In the future, we can improve the situation by adding detection of the database vendor and version and we would check this information before deciding to query or not. I suggest leaving a FIXME here.